### PR TITLE
adding solution to automate table snapshot creation

### PR DESCRIPTION
**Pull request by franklinWhaite**
This solutions automates and schedules the creation of [BigQuery Table Snapshots](https://cloud.google.com/bigquery/docs/table-snapshots-intro) at dataset and project level.

### CODE REVIEW

File: tools/cloud_functions/bq_table_snapshots/bq_backup_fetch_tables_names.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+from google.api_core import client_info as http_client_info
+import base64
+
+TABLE_TYPE_PHYSICAL_TABLE = "TABLE"
+# id of project used for BQ storage
+DATA_PROJECT_ID = ""
+# id of project with P/S topic
+PUBSUB_PROJECT_ID = ""
+# name of P/S topic where this code will publish to
+TABLE_NAME_PUBSUB_TOPIC_ID = "bq-backup-table-names"
```
Comment (danieldeleo) [inline tools/cloud_functions/bq_table_snapshots/bq_backup_fetch_tables_names.py:14]: Use environment variables https://cloud.google.com/functions/docs/configuring/env-var#setting_and_retrieving_runtime_environment_variables_an_example

File: tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py [hunk 2]
```diff
@@ -14,25 +15,36 @@
...
+    it = CronSim(cron_format, datetime.now())
+    next_interval = next(it)
+    next_next_interval = next(it)
+    delta = (next_next_interval - next_interval).total_seconds()
+    prev_interval = next_interval - relativedelta(seconds=delta)
+    prev_cron_interval_timestamp = int(prev_interval.timestamp() * 1000)
+    return prev_cron_interval_timestamp 
+
+
 def get_bq_client():
```
Comment (danieldeleo) [inline tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py:29]: Make the client a global variable to allow reuse across multiple invocations https://cloud.google.com/functions/docs/bestpractices/tips#use_global_variables_to_reuse_objects_in_future_invocations

File: tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    while True:
+        if job.done():
+            exception = job.exception()
+            if exception:
+                logging.info(str(exception))
+                raise Exception(str(exception))
+            else:
+                return 'ok'
+        time.sleep(2)
```
Comment (danieldeleo) [inline tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py:82]: add trailing newline

File: tools/cloud_functions/bq_table_snapshots/bq_backup_fetch_tables_names.py (hunk 1 above)
Comment (franklinWhaite) [reply tools/cloud_functions/bq_table_snapshots/bq_backup_fetch_tables_names.py]: done

File: tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py [hunk 4]
```diff
@@ -14,25 +15,36 @@
 BQ_JOBS_PROJECT_ID = ""
```
[the remark this answers is not in the document]
Comment (franklinWhaite) [reply tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py]: done

File: tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py (hunk 2 above)
Comment (franklinWhaite) [reply tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py]: done

File: tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py (hunk 3 above)
Comment (franklinWhaite) [reply tools/cloud_functions/bq_table_snapshots/bq_backup_create_snapshots.py]: done

**Review by danieldeleo — APPROVED** [review]

**Pull request merged**